### PR TITLE
Add `run_speech_recognition_seq2seq.py`

### PR DESCRIPTION
**Pull request by callumm-graphcore**
# What does this PR do?

Adds `run_speech_recognition_seq2seq.py` for training/fine-tuning Seq2Seq speech recognition models, such as Whisper, on the IPU.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by callumm-graphcore** [comment]
Sorry, should have clarified: this is no longer WIP

File: examples/speech-recognition/run_speech_recognition_seq2seq.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+        labels_batch = self.processor.tokenizer.pad(
+            label_features, pad_to_multiple_of=self.pad_to_multiple_of_labels, return_tensors="pt"
+        )
+
+        # replace padding with -100 to ignore loss correctly
+        labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)
+
+        # if bos token is appended in previous tokenization step,
+        # cut bos token here as it's append later anyways
+        if (labels[:, 0] == self.decoder_start_token_id).all().cpu().item():
```
Comment (katalinic-gc) [inline examples/speech-recognition/run_speech_recognition_seq2seq.py:267]: Wouldn't this step lead to labels of non-static shape?

File: examples/speech-recognition/run_speech_recognition_seq2seq.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+
+    processor = WhisperProcessor(feature_extractor, tokenizer)
+
+    # 10. Define data collator
+    data_collator = DataCollatorSpeechSeq2SeqWithPadding(
+        processor=processor,
+        decoder_start_token_id=model.config.decoder_start_token_id,
+        forward_attention_mask=forward_attention_mask,
+        # pad_to_multiple_of=math.ceil(max_input_length),
+        pad_to_multiple_of=80,
```
Comment (katalinic-gc) [inline examples/speech-recognition/run_speech_recognition_seq2seq.py:538]: If the processor is being set to Whisper, I don't believe this feature_extractor padding is necessary (the feature extractor will return examples of shape [1, 80, 3000].

File: examples/speech-recognition/run_speech_recognition_seq2seq.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+    # 11. Initialize Trainer
+    trainer = IPUSeq2SeqTrainer(
+        model=model,
+        ipu_config=ipu_config,
+        args=training_args,
+        train_dataset=vectorized_datasets["train"] if training_args.do_train else None,
+        eval_dataset=vectorized_datasets["eval"] if training_args.do_eval else None,
+        data_collator=data_collator,
+        compute_metrics=compute_metrics if training_args.predict_with_generate else None,
+        eval_parallelize_kwargs={
```
Comment (katalinic-gc) [inline examples/speech-recognition/run_speech_recognition_seq2seq.py:551]: With https://github.com/huggingface/optimum-graphcore/pull/427 merged, this should instead be passed to the IPUConfig as a `inference_parallelize_kwargs` entry.

**Comment by katalinic-gc** [comment]
The failing test is the one that compares the diff file. Is the current one up to date?

**Comment by callumm-graphcore** [comment]
> The failing test is the one that compares the diff file. Is the current one up to date?
I thought it was, I'll double check
